### ISSUE TITLE
installation under MinGW/MSYS does not work

### ISSUE THREAD

**Issue by arjan64** [sourceryinstitute/opencoarrays#40]
Hi!

I tried to let the script make an installation under MinGW/MSYS, but it did not work. The first crashes made clear that I should add wget and svn. After these had been installed, I re-ran the script, the runlog of which is attached to this message. Somehow, my gcc compiler was not approved of, with a reference to a file config.log with details. Unfortunately, this platform does not allow for uploading files of type .log, so I renamed it to config.txt. The config.log that I found in subdirectory gcc-trunk-build was renamed to config_2.txt and also attached. Please tell me how to proceed!

Regards,

Arjan

[runlog.txt](https://github.com/sourceryinstitute/opencoarrays/files/52798/runlog.txt)
[config.txt](https://github.com/sourceryinstitute/opencoarrays/files/52799/config.txt)
[config_2.txt](https://github.com/sourceryinstitute/opencoarrays/files/52803/config_2.txt)

**Comment by sourceryinstitute** [comment]
Hi Arjan,

When I have seen this before, it's usually because the the CC variable didn't get set correctly by the script.  Are you able to set the CC  environment variables to a working C compiler of your choice and ensure that the environment variable gets passed to the install.sh script?  In a bash shell, for example, you might do so with the gcc compiler by invoking the script as follows:

CC=gcc ./install.sh

or, if the compiler is not in your path, then

CC=/absolute/path/goes/here/gcc ./install.sh

Here's a bit of explanation that might help.  install.sh works its way recursively down the following dependency stack looking to find or install each package and progressing to the next lower package when it cannot find a given package and will therefore need to install one or more prerequisites for building the missing package:

mpif90 wrapping gfortran 5.1.0 or later
gfortran 5.1.0 or later 
flex
bison

From the 'runlog.txt' file, it appears that you probably restarted install.sh after a previous run (which is run), the install.sh script found the flex source tar ball that was a downloaded during the previous run, and you approved using it to build flex.  In order to reach the flex build, install.sh has to first complete its search for an mpif90 wrapping gfortran 5.1.0 or later.  If that isn't found, it searches for gfortran 5.1.0 or later itself (which might be present even if mpif90 doesn't wrap it).  If that isn't found, the script searches  for flex.  In your case it finds the flex source, at which point it prepares to build flex by setting the CC=gcc variable if CC is empty.  If CC is not empty, then it leaves the CC value untouched.  Therefore, if you're certain that you have gcc on your system, then hopefully you can fix this problem by setting CC and _ensuring_ that the CC value get passed to the subshell that launches when the script launches.  The aforementioned syntax

CC=gcc ./install.sh

ensures that the environment variable passes to a subshell in bash.  You might need another way to ensure the passing if you're using a different shell.

Unfortunately, I am going to be extremely pressed for time over the next several days so it might take me a day or two to respond to issues this week.

**Comment by arjan64** [comment]
Hi!
Thanks for the suggestions, but my agenda is pretty crammed either, so I will check what you wrote, but only in a few days.
Cheers,

Arjan

**Comment by zbeekman** [comment]
It does look like the issue is finding the C compiler, as Damian mentioned above. You may also run into #81 if get a little bit further, but of course we'd love to hear otherwise...

**Comment by SourcerersApprentice** [comment]
The compiler version (2.95, built in 1999) looks like it might be a little weak. Is there a minimumgcc version that is able to build gfortran?

**Comment by SourcerersApprentice** [comment]
I would recommend mingw-get-setup.exe from http://sourceforge.net/projects/mingw/files/ to get a more recent gcc build compiler and environment. Then, open up a bash shell and try the install again.

**Comment by arjan64** [comment]
Today, I deleted all old attempts and directories and went to the suggested http://sourceforge.net/projects/mingw/files/ for a fresh basic gcc and performed the installation. Then I re-downloaded wget.exe, placed it in C:\MinGW\msys\1.0\bin, and started the install script for opencoarrays as suggested earlier as follows:

CC=gcc ./build.sh flex >& runlog_flex.txt

It crashed, see attachments (config.log was renamed to config.txt, since this site does not want to upload .log-files...). 

New suggestions?


[runlog_flex.txt](https://github.com/sourceryinstitute/opencoarrays/files/74066/runlog_flex.txt)
[config.txt](https://github.com/sourceryinstitute/opencoarrays/files/74067/config.txt)

**Comment by rouson** [comment]
It appears the problem is with the pre-installed C compiler.  config.txt contains

gcc.exe: installation problem, cannot exec `cc1': No such file or directory

Please let us know if this [page](http://stackoverflow.com/questions/11912878/gcc-error-gcc-error-trying-to-exec-cc1-execvp-no-such-file-or-directory) helps.

**Comment by SourcerersApprentice** [comment]
I agree it looks like you have not properly installed the new compiler, it still is using the old one. Make sure you are using the msys shell that you recently downloaded, it has the paths set to the right compiler and tools. You want to see a compiler version that looks like 4.8 or 4.9, typing "gcc -v" will display the version.

**Comment by arjan64** [comment]
I was sure that I had erased any old gcc stuff before I made my new attempt today, including the desktop icon to the MSYS shell. Therefore the old version number of gcc puzzled me. I typed "which gcc" and got /c/FPC/2.6.2/bin/i386-Win32/gcc.exe, so it appears that gcc is first found in my ancient FreePascal installation! Now I am re-trying my opencoarrays installation with explicit path specified for gcc. Keep fingers crossed!

**Comment by arjan64** [comment]
The log now tells me that it does use gcc version 4.8.1, but installation still crashes, see attachments:

configure:2965: /c/MinGW/bin/gcc -V >&5
gcc.exe: error: unrecognized command line option '-V'
gcc.exe: fatal error: no input files
compilation terminated.

How to proceed?

[config.txt](https://github.com/sourceryinstitute/opencoarrays/files/74356/config.txt)
[runlog_flex.txt](https://github.com/sourceryinstitute/opencoarrays/files/74357/runlog_flex.txt)

**Comment by SourcerersApprentice** [comment]
This might be an install script problem, it looks like runtime headers are not installed.
For a Windows build using Mingw, the Mingw runtime needs to be downloaded and the headers installed to a directory under the sysroot named "<sysroot>/mingw/include" before starting the configure. This library is largely a thin interface to the Microsoft C runtime DLL "msvcrt". After the compiler is built, but before it starts to build its language runtime libraries, the mingw runtime needs to be built and installed to a directory named "<sysroot>/mingw/lib". The Windows pthreads implementation is included with the mingw runtime, and headers and libraries should be built/installed like the main runtime.
Since the install uses svn, that makes getting the runtime easy at 
svn://svn.code.sf.net/p/mingw-w64/code/trunk
It uses the normal "configure" style build.

**Comment by rouson** [comment]
@SourcerersApprentice, from what you've written above, I assume the "install script problem" you're referencing is not an OpenCoarrays installation script (i.e., not "install.sh" or "build").   If I'm incorrect and there is a problem with an OpenCoarrays installation script, please let me know.

**Comment by SourcerersApprentice** [comment]
If the OpenCoarrays install script is meant to support Mingw compiler builds, then there is a problem with the script. The extra runtime pieces described above should be considered as a prerequisite for the compiler, and are a substitution for libc. If the install script was only intended to support Cygwin environments (this sort of seems implied), then what we are trying is experimental. We know that we can't take the coarrays build all the way yet with mingw, but there are no high hurdles to jump getting the gfortran build part working. If you want to add this ability, I am willing to help out when you hit any sticky patches.

**Comment by rouson** [comment]
Ideally OpenCoarrays would support every common platform on which one expects users to write CAF programs.  That certainly includes Windows.  I have no strong opinion about whether it includes Cygwin, Mingw or any other Windows environment only because I haven't used any of them in years and I wrote the installation script.  If those who use Windows regularly suggest that Mingw should be supported, then I would gladly incorporate any edits to the script required to support it.  The biggest problem we face is simply access and expertise.  If someone would like to contribute to improving the script, I would be glad to walk anyone interested through the details of the script to help out.  Then it would be best if one of you forks the repository, creates a new branch for madden the edits, and submits a pull request.

**Comment by SourcerersApprentice** [comment]
My scripting skills aren't really strong, but I do know how to do the Windows build, so I am willing to give it a try. I won't be able to get to it until mid-next week, though.

**Comment by arjan64** [comment]
> For a Windows build using Mingw, the Mingw runtime needs to be downloaded 
> and the headers installed [...] svn://svn.code.sf.net/p/mingw-w64/code/trunk

I don't have the 64-bits version of Windows-7 installed, though that would have been possible on this box. Therefore, I fear that the svn link given above may not lead to something working on my 32-bit Windows-7 machine. Furthermore, my experience with svn is nil. Can I instead amend my MinGW/MSYS installation with the "MinGW Installation Manager"? That tool is rather cumbersome, so some assistance in which items to select is welcome. I will now proceed and activate all stuff under "Basic Setup" (I had most of it installed, but not yet everything) and then re-try the opencoarrays installation. 

I do agree that it would be nice that if an OpenCoarray install were to support MinGW/MSYS, then this type of handlings would in the future preferably be caught and dealt with by the script and not by the user. I hope that someone capable of updating that script, and in the position to put some time in it ;-)) , will stand up and put this a step further. And there is no rush, I just want to use the new SUBMODULES feature, but otherwise my code runs fine, so mid-next week is okay for me!

**Comment by rouson** [comment]
My preference would be to work on this sometime after January 19.  Please feel free to book a time at http://rouson.youcanbook.me.  My bash scripting skills were non-existent before writing the install.sh.  I picked up everything I needed to know by the online search-copy-modify method.  Let's plan to talk through what is in the script and what needs to be done.

**Comment by JulioBeneyto** [comment]
I have caf.c . You can compile coarrays. The method for compiling is : gcc -c -o caf.o caf.c. A fortran program is  compiled with : gfortran -cpp -gfortran=lib nameOfProgram.f90 cof.o. 
I tested the OpenCoarrays porgrams: allocate_as_barrier.f90 Test passed. co_broadcast.f90 Test passed. hello_multiverse.f90 Test passed. syncall.f90 Test passed. syncimages.f90 Test passed. syncimages2.f90 Test passed. test1.caf Test passed. whole_get_array.f90m Test 5 fails. 
If you run with a.exe the program uses NumberOfProcessor of Windows. else a.exe -n 2 the program uses 2 processors.

**Comment by JulioBeneyto** [comment]
I am sorry. gfortran -cpp -fcoarray=lib nameOfProgram.f90 cof.o

**Comment by zbeekman** [comment]
@JulioBeneyto I'm glad to hear you were able to compile and run OpenCoarrays. However, could you please give us some additional details about your system and setup? Also, it appears that you may be leaving something out, since I see no linking happening with an MPI implementation.

 Please include the following details:
  - OS and version
  - Linux emulation environment and version (i.e., Cygwin or MinGW/MSYS)
  - MPI implementation and version
  - All additional linking and compiling commands and their outputs

Thanks!

**Comment by JulioBeneyto** [comment]
Hi Izaak
1.- Windows 7 64 bits
2.- Cygwin & MinGW/MSYS
3.- MPI implementation don't have.
4.- First gcc -c -o caf.o caf.c and their outputs is 'caf.o'
     Finally you compile gfortran -cpp -fcoarray=lib NameOfProgram.f90 
caf.o to run ./a or ./a -n (2,3,..,32)
Please see the file attached caf.zip
Thanks Julio

**Comment by zbeekman** [comment]
@JulioBeneyto I'm still confused, why would  you want to build OpenCoarrays *without* MPI? Also, what is the file `caf.c` you mentioned? The main OpenCoarrays implementation is in https://github.com/sourceryinstitute/opencoarrays/blob/master/src/mpi/mpi_caf.c

I'm guessing that what you're seeing is gfortran's single processor support for CAF? Maybe @rouson or @fanfa can weigh in here.

For example, what happens when you run this program on 4 processors:

```fortran
      program main
        use iso_c_binding, only : c_int
        use iso_fortran_env, only : error_unit
        implicit none
        integer(c_int) :: tally
        tally = this_image() ! this image's contribution
        call co_sum(tally)
        verify: block
          integer(c_int) :: image
          if (tally/=sum([(image,image=1,num_images())])) then
             write(error_unit,'(a,i5)') "Incorrect tally on image ",this_image()
             error stop 2
          end if
        end block verify
        ! Wait for all images to pass the test
        sync all
        if (this_image()==1) write(*,*) "Test passed"
      end program
```

**Comment by JulioBeneyto** [comment]
Hi Izaak
Because OpenMPI of Cygwin and MinGW don't support multi-process.($ 
ompi_info the answer is MPI_THREAD_MULTIPLE: NO)
caf.c is my mpi_caf.c, it uses CreateProcess (win32 API) to create 
multi-process. The information,it's sent to other process via shared 
memory.
You program main.f90 is 0K with:
integer(c_int) :: tally[*]
If you compile with
gfortran -cpp -fcoarray=lib main.f90 caf.o
./a
    Test passed
./a -n 4
    Test passed
Bye

**Comment by zbeekman** [comment]
Hi all, this thread has been dormant for quite some time. If you really want to use OpenCoarrays on Windows, your best bet is probably some sort of Linux virtualization. I highly recommend vagrant (www.vagrantup.com) with VirtualBox. It currently sounds as though the issues are on the mingw/cygwin side, and neither platform is officially supported at this time. If you have made any more attempts to get this working, we'd certainly love to hear about them, but for now I'm going to mark this issue as "won't fix" and close it due to inactivity and no official support or means for us to reproduce the issue.
